feat: operator creates app and UI Ingresses

jsenko opened the pull request.

jsenko commented on operator/controller/src/main/java/io/apicurio/registry/operator/resource/ui/UIIngressDiscriminator.java:
```diff
@@ -0,0 +1,24 @@
+
+import static io.apicurio.registry.operator.resource.ResourceFactory.COMPONENT_UI;
+
+public class UIIngressDiscriminator extends LabelDiscriminator<Ingress> {
```
Sure, I'll do that

jsenko commented on operator/controller/src/main/java/io/apicurio/registry/operator/util/BeanUtil.java:
```diff
@@ -0,0 +1,38 @@
+import java.util.function.Function;
+
+/**
+ * This is needed because QOSDK does not support bean injection into dependent resource classes. We work
```
There was a comment I found, but I can't find it anymore. I have confirmed it experimentally though. We can use singletons in this case, but when I update the code with more features, the problem will be back. We can use singletons despite this, but I expect it'll be much more inconvenient.

andreaTP replied:
Is it this one?
https://github.com/quarkiverse/quarkus-operator-sdk/issues/720

andreaTP replied:
In that case I'd suggest trying to avoid using DI even if it would be convenient, and rely on configuration and generic/static methods.

Let see how far we can go with this approach, if it's enough for other operators I assume it would be fine for us despite it being sub-optimal.

jsenko replied:
If you insist, but I think it's worth it using CDI.

andreaTP replied:
In general, I would try to go in a direction where the discriminators are "simpler", wondering how other projects are doing ( [KC](https://github.com/search?q=repo%3Akeycloak%2Fkeycloak%20Discriminator&type=code) )

jsenko commented on operator/controller/src/main/java/io/apicurio/registry/operator/util/HostUtil.java:
```diff
@@ -0,0 +1,29 @@
+import org.slf4j.LoggerFactory;
+
+@ApplicationScoped
+public class HostUtil {
```
We can now, but I'll inject beans into this next, so I'll be just reverting it.

jsenko commented on operator/controller/src/main/java/io/apicurio/registry/operator/util/HostUtil.java:
```diff
@@ -0,0 +1,29 @@
+
+    public String getHost(String component, ApicurioRegistry3 p) {
+        var prefix = p.getMetadata().getName() + "-" + component + "." + p.getMetadata().getNamespace();
+        String host = prefix + ".cluster.example";
```
Just a dummy value. I'll add a field in the CRD so users can configure it and also detection of canonical route hostname on OpenShift.

andreaTP replied:
IIUC ".cluster.example" should be configurable from the `.properties` file.

andreaTP replied:
> I'll add a field in the CRD so users can configure it and also detection of canonical route hostname on OpenShift.

I agree that the user should be able to configure it, and I think that automatic detection is out of scope for now.

jsenko replied:
Why not? Our current operator does this, if we don't add it, users (including us developing it) would have to manually add the CRH every time we develop an test the operator on OCP. It's not a big feature, and we will add it later anyway, so why delaying the convenience?

andreaTP replied:
What are other operators doing for this case?
I have not seen automatic detection implemented so far.

jsenko replied:
I don't see a strong argument here. I could go investigating other operators for a few hours until I find something but this is a convenience feature that does not cost us anything, [has been requested by users](https://github.com/Apicurio/apicurio-registry-operator/issues/19), has worked well in the legacy operator and is not complext to implement.

andreaTP replied:
> but this is a convenience feature that does not cost us anything

We should strive to keep features to a minimum and avoid doing assumptions as much as possible.
In this case, auto-detection means that the operator should run in the same cluster and possibly with the same network rules of the operand.
In a lot of case, this doesn't hold true.

At the same time, adding a configuration property is simpler, more straightforward and with much less code to be updated/maintained, giving the user the functionality.

But this discussion is not relevant to this PR.

jsenko commented on operator/controller/src/main/java/io/apicurio/registry/operator/resource/ui/UIDeploymentResource.java:
```diff
@@ -31,16 +37,33 @@ protected Deployment desired(ApicurioRegistry3 primary, Context<ApicurioRegistry
+            var iOpt = context.getSecondaryResource(APP_INGRESS_KEY.getKlass(),
+                    APP_INGRESS_KEY.getDiscriminator());
+            iOpt.ifPresent(i -> {
+                for (IngressRule rule : i.getSpec().getRules()) {
```
ok

andreaTP commented on operator/controller/src/test/java/io/apicurio/registry/operator/it/SmokeITTest.java:
```diff
@@ -40,5 +44,16 @@ void demoDeployment() {
         });
+
+        // Ingresses
+        await().ignoreExceptions().until(() -> {
```
Since now, we have access to both the API and the UI, can you please, add an integration test to access those?

andreaTP commented on operator/controller/src/main/java/io/apicurio/registry/operator/util/IngressUtil.java:
```diff
@@ -0,0 +1,34 @@
+
+    public static String getHost(String component, ApicurioRegistry3 p) {
+        var prefix = p.getMetadata().getName() + "-" + component + "." + p.getMetadata().getNamespace();
+        String host = prefix + ".cluster.example";
```
Please, get the suffix from the CR with a default to `.cluster.local` I think.

jsenko replied:
I've seen that domain being used, but it seems like it's reserved for another purpose (Multicast DNS) https://en.wikipedia.org/wiki/Special-use_domain_name .

jsenko replied:
Is that applicable here?

andreaTP replied:
This is the hostname as seen from outside of the cluster, correct?
Would it work with a blank suffix to start with?

jsenko replied:
yes. Sure I can do that.

andreaTP commented:
Forgot to mention, we also need a top level CR field to disable the default ingress creation for the case where someone wants a highly customized one.

It would be nice in this PR, but can also be done as a follow up. :+1:

jsenko commented:
> Forgot to mention, we also need a top level CR field to disable the default ingress creation for the case where someone wants a highly customized one.
> 
> It would be nice in this PR, but can also be done as a follow up. 👍

Agreed, I'll do it in the next PR.

andreaTP commented on operator/controller/src/main/java/io/apicurio/registry/operator/util/Util.java:
```diff
@@ -0,0 +1,11 @@
+    private Util() {
+    }
+
+    public static boolean isBlank(String value) {
```
nitpick: this is used exclusively in `IngressUtil`, move this method to be private in the using class.

andreaTP commented on operator/controller/src/main/java/io/apicurio/registry/operator/util/IngressUtil.java:
```diff
@@ -0,0 +1,60 @@
+
+    public static String getHost(String component, ApicurioRegistry3 p) {
+        String host = null;
+        if (COMPONENT_APP.equals(component)) {
```
I'm not convinced about this refactoring to a `Util` class.
The fact that we need to match on strings to customize the logic is a strong indication that this is not the right level of abstraction.
The rule of thumb for "Util", "Common" and similar classes is:
> if you can't name it properly it should probably not exists

I propose that we remove this `IngressUtil` class, if the cost is some code duplication, I'm ok with it for now.

jsenko replied:
We can have a separate session later to discuss improving the code structure.

andreaTP replied:
ok, deal

andreaTP commented on operator/model/src/main/java/io/apicurio/registry/operator/api/v1/ApicurioRegistry3Spec.java:
```diff
@@ -1,17 +1,36 @@
+     * Apicurio Registry backend base URL
+     */
+    @JsonProperty("appHost")
+    @JsonPropertyDescription("Apicurio Registry backend base URL")
```
base ?

jsenko replied:
Meaning that it's not URL directly to the API, so it's not directly usable in e.g. serdes.

andreaTP replied:
oh, ok ... not sure what would be a better name though

andreaTP reviewed (commented):
This is almost ready, just missing a few tests accessing the instances through the `Ingress`es.

Good job! :+1:

jsenko commented:
> This is almost ready, just missing a few tests accessing the instances through the `Ingress`es.
> 
> Good job! 👍

Thank you! I'm working on the tests ATM.

jsenko commented:
@andreaTP I've added tests that make requests to Registry via Service (port-forwarding) and Ingress (minikube tunnel).

I've also refactored the tests a bit (same logic). The unit tests are now always executed as local tests, and integration tests are now always executed as remote tests (test logic is shared). The reason is that when remote tests are executed as Quarkus unit tests, local operator is always deployed, even if we test against an instance in the cluster. Now these concerns are separated and configuration is simpler. Please read the [README.md file](https://github.com/Apicurio/apicurio-registry/blob/1d419ee0e5776ec6615bc41762b9e103e0bcb25d/operator/controller/README.md).

There's a non-blocker problem in the workflows with maven logs being too verbose, if you have an idea about the cause, any help is appreciated.

jsenko commented:
@andreaTP review please

andreaTP commented on .github/workflows/operator.yaml:
```diff
@@ -28,13 +28,27 @@ jobs:
+
+      - name: Enable port-forwarding
+        run: |
+          sudo apt-get -y install socat  
```
nitpick: is `socat` needed? the tunnel should already do the job.

jsenko replied:
yes it's needed

andreaTP commented on operator/controller/src/test/java/io/apicurio/registry/operator/it/SmokeITTest.java:
```diff
@@ -40,5 +53,99 @@ void demoDeployment() {
+
+    @Test
+    void testIngress() {
+        if (ingressManager.isIngressSupported()) {
```
nitpick: an idiomatic way of doing this would be to use something like: `@EnabledIfSystemProperty` or similar from JUnit

andreaTP reviewed (approved):
A couple of nitpicks but LGTM :+1:
